Objects/unicodeobject.c

```diff
@@ -11343,11 +11349,49 @@ Return -1 on failure.
 [clinic start generated code]*/
 
 static Py_ssize_t
-unicode_find_impl(PyObject *str, PyObject *substr, Py_ssize_t start,
+unicode_find_impl(PyObject *str, PyObject *subobj, Py_ssize_t start,
                   Py_ssize_t end)
-/*[clinic end generated code: output=51dbe6255712e278 input=4a89d2d68ef57256]*/
+/*[clinic end generated code: output=80175735a6d549d0 input=51e7b530950ab304]*/
 {
-    Py_ssize_t result = any_find_slice(str, substr, start, end, 1);
+    Py_ssize_t result;
+    if (PyTuple_Check(subobj)) {
+        result = -1;
+        Py_ssize_t len = PyUnicode_GET_LENGTH(str);
+        ADJUST_INDICES(start, end, len);
+        // Work in batches of 10000
+        for (; result == -1 && start <= end; start += 10000) {
```

Comment:
Should it be `<`? If yes, same for bytes?

Reply:
I think this is bad optimization.
How this batch improve maximum/typical complexity?

Reply:
Depends what to compare to. The same method without batches has no early stopping and runs full scan for each substring. For substrings of length 1 complexity improvement is:
Best case: `O(n * nsubs) -> O(min(10K, n) * nsubs)`
Worst case: Complexity does not change

For large strings with hits at the back there is an overhead of extra C function calls, but overhead is acceptable even for large strings. Tested for length of 1M. Without this optimisation it becomes highly suboptimal for cases of long strings with an early match compared to other methods.

Python function call with all preparation takes approximately as long as 10K scan so in practice this whole approach performs well. This solution pretty much ends up at least as fast and often faster compared to any other existing method. While performance of other methods generally varies from case to case: how early the hit is, is it one off case or pre-compiling is amortized.

If the functionality is acceptable, specialised algorithm can always be added. But I don't think it would lead to any worthwhile performance gain (except some extreme corner cases that I have never seen in reality). `str.find` is an efficient method and this seems to manage to take a good advantage of that.